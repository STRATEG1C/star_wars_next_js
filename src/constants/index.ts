export const API_URL = 'http://swapi.dev';

export const API_PEOPLE = `${API_URL}/api/people/`;

export const API_RANDOM_IMAGE = 'https://picsum.photos';

export const API_RANDOM_PEOPLE = 'https://randomuser.me/api';

export const DEFAULT_NUMBER_OF_POSTS = 6;
