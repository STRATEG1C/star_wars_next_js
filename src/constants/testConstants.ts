import { IPost } from '../components/PostsList/types';

export const testPost: IPost = {
    thumbnail: '',
    author: {
        avatar: 'string',
        name: 'Dart Veider',
        specie: 'Human',
        home: 'Star'
    }
};

export const testPostList: IPost[] = [
    {
        thumbnail: '',
        author: {
            avatar: 'string',
            name: 'Dart Veider',
            specie: 'Human',
            home: 'Star'
        }
    },
    {
        thumbnail: '',
        author: {
            avatar: 'string',
            name: 'Obivan',
            specie: 'Human',
            home: 'Star'
        }
    }
];
