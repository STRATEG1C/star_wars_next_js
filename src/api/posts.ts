import axios, { AxiosResponse } from 'axios';
import { API_PEOPLE, API_RANDOM_PEOPLE } from '../constants';
import { ICharacter, IHomeworld, IListResponse, ISpecie } from '../types';
import { IPost } from '../components/PostsList/types';

export const loadPosts = async (limit): Promise<IPost[]> => {
  const charactersRes = await axios.get(API_PEOPLE) as AxiosResponse<IListResponse<ICharacter[]>>;

  const characters = charactersRes.data.results;

  const randomUserRes = await axios.get(API_RANDOM_PEOPLE) as AxiosResponse;

  const postList: IPost[] = await Promise.all(characters.map(async (character: ICharacter) => {
    let userImage = '';
    let characterSpecie = 'Unknown';
    let characterHomeworld = '';

    const characterHomeworldRes = await axios.get(character.homeworld) as AxiosResponse<IHomeworld>;

    userImage = randomUserRes.data.results[0].picture.medium;
    characterHomeworld = characterHomeworldRes.data.name

    if (character.species[0]) {
      const characterSpecieRes = await axios.get(character.species[0]) as AxiosResponse<ISpecie>
      characterSpecie = characterSpecieRes.data.name;
    }

    return {
      thumbnail: '',
      author: {
        avatar: userImage,
        name: character.name,
        specie: characterSpecie,
        home: characterHomeworld
      }
    };
  }));

  return postList.splice(0, limit);
}
