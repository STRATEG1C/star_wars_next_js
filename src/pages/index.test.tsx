import React from 'react';
import { shallow, render } from 'enzyme';

import { AppContext } from '../context';
import { DEFAULT_NUMBER_OF_POSTS } from '../constants';
import { loadPosts } from '../api/posts';

import Home from './index';

describe('Test of main page', () => {
    const loadPostsMock = jest.fn();

    it('Should render the page', () => {
        const wrapper = shallow(
            <AppContext.Provider value={{
                posts: [],
                resetPosts: loadPostsMock()
            }}>
                <Home postList={[]} />
            </AppContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('Should render the page with received posts', async () => {
        const postList = await loadPosts(DEFAULT_NUMBER_OF_POSTS);

        const wrapper = render(
            <Home postList={postList} />
        );

        expect(wrapper).toMatchSnapshot();
    });
})
