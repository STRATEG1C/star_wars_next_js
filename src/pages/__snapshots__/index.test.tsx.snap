// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of main page Should render the page 1`] = `
<Home
  postList={Array []}
/>
`;

exports[`Test of main page Should render the page with received posts 1`] = `
<div>
  <header
    class="header"
  >
    <h2
      class="font-bold text-3xl"
    >
      The Fight Continues
    </h2>
    <div>
      <span
        class="font-bold"
      >
        0 Warriors
      </span>
      <button
        class="btn btn-blue ml-8 w-32"
      >
        Reset
      </button>
    </div>
  </header>
  <main>
    <p
      class="mt-10 ml-5"
    >
      <label
        for="posts-quantity"
      >
        Posts quantity:Â 
      </label>
      <select
        name="posts-quantity"
      >
        <option
          value="3"
        >
          3
        </option>
        <option
          selected=""
          value="6"
        >
          6
        </option>
        <option
          value="9"
        >
          9
        </option>
      </select>
    </p>
    <div
      class="mt-10"
    >
      <div
        class="post-list flex flex-wrap"
      />
    </div>
  </main>
</div>
`;
